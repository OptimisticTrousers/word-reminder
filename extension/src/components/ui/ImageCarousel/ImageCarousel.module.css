.carousel {
  max-width: 1000px;
  position: relative;
  margin: auto;
}

.carousel__dot {
  border-radius: 12px;
}

.carousel__image {
  width: 100%;
}

button.carousel__button--prev,
button.carousel__button--next {
  position: absolute;
  top: 50%;
  width: auto;
  margin-top: -22px;
  padding: 16px;
  transition: 0.6s ease;
  border-radius: 0 3px 3px 0;
  user-select: none;
}

.carousel__button--prev {
  left: 0;
}

.carousel__button--next {
  right: 0;
}

.carousel__scroll {
  margin-top: 24px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.carousel__caption {
  color: #f2f2f2;
  font-size: 15px;
  padding: 8px 12px;
  width: 100%;
  text-align: center;
}

.carousel__count {
  color: var(--black-text-color);
  font-size: 12px;
  z-index: 1;
  padding: 8px 12px;
  position: absolute;
  background-color: var(--black-bkg-color);
}

.carousel__buttons {
  margin-top: 8px;
}

.carousel__button--dot {
  height: 30px;
  width: 30px;
  margin: 0 2px;
  margin-right: 15px;
  border-radius: 50%;
  display: inline-block;
  transition: background-color 0.6s ease;
}

button.carousel__button--active {
  background-color: var(--secondary-icon-color);
}

.fade {
  animation-name: fade;
  animation-duration: 1.52;
}

.carousel__dots {
  display: flex;
  justify-content: center;
  align-items: center;
}

@keyframes fade {
  from {
    opacity: 0.4;
  }
  to {
    opacity: 1;
  }
}
