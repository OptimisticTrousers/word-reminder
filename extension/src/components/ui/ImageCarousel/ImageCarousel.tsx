import { ChevronLeft, ChevronRight } from "lucide-react";
import { useState } from "react";
import CSSModules from "react-css-modules";

import styles from "./ImageCarousel.module.css";
import { useInterval } from "../../../hooks/useInterval";
import { Image } from "common";
import noImageAvailableImage from "../../../assets/images/no-image-available.jpg";

export interface Props {
  images: Image[];
  hasAutoScroll: boolean;
}

export const ImageCarousel = CSSModules(
  function ({ images, hasAutoScroll }: Props) {
    const [slideIndex, setSlideIndex] = useState(1);

    const delay = 500;
    useInterval(
      () => {
        plusSlides(1);
      },
      hasAutoScroll ? delay : null
    );

    function plusSlides(number: number) {
      setSlideIndex((prevSlideIndex) => {
        const newIndex = prevSlideIndex + number;
        if (newIndex > images.length) {
          return 1;
        } else if (newIndex < 1) {
          return images.length;
        }
        return newIndex;
      });
    }

    function currentSlide(index: number) {
      setSlideIndex(index);
    }

    return (
      <div styleName="carousel">
        <div styleName="carousel__images">
          {images.length > 0 ? (
            images.map((image: Image, index: number) => {
              const position = index + 1;
              return (
                position === slideIndex && (
                  <figure key={index} styleName="carousel__item fade">
                    <span styleName="carousel__count">
                      {position}/{images.length}
                    </span>
                    <img
                      styleName="carousel__image"
                      src={image.src}
                      alt=""
                      role="presentation"
                    />
                    <figcaption styleName="carousel__caption">
                      {image.caption}
                    </figcaption>
                  </figure>
                )
              );
            })
          ) : (
            <figure styleName="carousel__item fade">
              <span styleName="carousel__count">0/0</span>
              <img
                styleName="carousel__image"
                src={noImageAvailableImage}
                alt=""
                role="presentation"
              />
              <figcaption styleName="carousel__caption">
                No Images Available
              </figcaption>
            </figure>
          )}
          <button
            styleName="carousel__button carousel__button--prev"
            aria-label="Previous image"
            onClick={() => {
              plusSlides(-1);
            }}
          >
            <ChevronLeft styleName="carousel__icon" />
          </button>
          <button
            styleName="carousel__button carousel__button--next"
            aria-label="Next image"
            onClick={() => {
              plusSlides(1);
            }}
          >
            <ChevronRight styleName="carousel__icon" />
          </button>
        </div>
        <div styleName="carousel__dots">
          {images.map((_, index: number) => {
            const position = index + 1;
            return (
              <button
                key={index}
                styleName={`carousel__button carousel__button--dot ${
                  slideIndex === position && "carousel__button--active"
                }`}
                aria-label={`Show image ${position}`}
                onClick={() => currentSlide(position)}
              ></button>
            );
          })}
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
