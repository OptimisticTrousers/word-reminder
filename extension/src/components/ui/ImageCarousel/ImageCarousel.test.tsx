import { act, render, screen } from "@testing-library/react";

import laRocherDeBaumeImage from "../../../assets/test/la-rocher-de-baume.jpg";
import legoImage from "../../../assets/test/lego.jpg";
import selfParkingSignImage from "../../../assets/test/self-parking-sign.jpg";
import { ImageCarousel } from "./ImageCarousel";
import userEvent from "@testing-library/user-event";

describe("ImageCarousel component", () => {
  const images = [
    { src: laRocherDeBaumeImage, caption: "La Rocher De La Baume" },
    { src: legoImage, caption: "Crowd Lego Staff" },
    { src: selfParkingSignImage, caption: "UFO Parking Sign" },
  ];

  it("shows the first image by default", () => {
    const { asFragment } = render(
      <ImageCarousel images={images} hasAutoScroll={false} />
    );

    const count = screen.getByText("1/3");
    const image = screen.getByRole("presentation");
    const caption = screen.getByText(images[0].caption);
    expect(count).toBeInTheDocument();
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute(
      "src",
      "/src/assets/test/la-rocher-de-baume.jpg"
    );
    expect(caption).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows no images placeholder when there are no images", () => {
    const { asFragment } = render(
      <ImageCarousel images={[]} hasAutoScroll={false} />
    );

    const count = screen.getByText("0/0");
    const image = screen.getByRole("presentation");
    const caption = screen.getByText("No Images Available");
    expect(count).toBeInTheDocument();
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute(
      "src",
      "/src/assets/images/no-image-available.jpg"
    );
    expect(caption).toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  });

  describe("hasAutoScroll", () => {
    const delay = 5000;

    beforeEach(() => {
      vi.useFakeTimers();
    });

    afterEach(() => {
      vi.runOnlyPendingTimers();
      vi.useRealTimers();
    });

    it("it scrolls to the next image in 5 seconds", async () => {
      const { unmount } = render(
        <ImageCarousel images={images} hasAutoScroll={true} />
      );
      act(() => {
        vi.advanceTimersByTime(delay);
      });

      const count = screen.getByText("2/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[1].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(caption).toBeInTheDocument();
      // Prevents act() error caused by the useInterval hook continuing to run after the test finishes
      unmount();
    });

    it("it scrolls to to the last image in 10 seconds", () => {
      const { unmount } = render(
        <ImageCarousel images={images} hasAutoScroll={true} />
      );
      act(() => {
        vi.advanceTimersByTime(delay);
        vi.advanceTimersByTime(delay);
      });

      const count = screen.getByText("3/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[2].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(caption).toBeInTheDocument();
      // Prevents act() error caused by the useInterval hook continuing to run after the test finishes
      unmount();
    });

    it("it scrolls to the first image in 15 seconds", () => {
      const { unmount } = render(
        <ImageCarousel images={images} hasAutoScroll={true} />
      );
      act(() => {
        vi.advanceTimersByTime(delay);
        vi.advanceTimersByTime(delay);
        vi.advanceTimersByTime(delay);
      });

      const count = screen.getByText("1/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[0].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(caption).toBeInTheDocument();
      // Prevents act() error caused by the useInterval hook continuing to run after the test finishes
      unmount();
    });
  });

  describe("clicking previous", () => {
    it("goes to the previous image", async () => {
      render(<ImageCarousel images={images} hasAutoScroll={false} />);
      const user = userEvent.setup();
      const thirdImageDot = screen.getByRole("button", {
        name: "Show image 3",
      });
      await user.click(thirdImageDot);

      const previousButton = screen.getByRole("button", {
        name: "Previous image",
      });
      await user.click(previousButton);

      const count = screen.getByText("2/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[1].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute("src", "/src/assets/test/lego.jpg");
      expect(caption).toBeInTheDocument();
    });

    it("goes to the last image when on the first image", async () => {
      render(<ImageCarousel images={images} hasAutoScroll={false} />);
      const user = userEvent.setup();

      const previousButton = screen.getByRole("button", {
        name: "Previous image",
      });
      await user.click(previousButton);

      const count = screen.getByText("3/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[2].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute(
        "src",
        "/src/assets/test/self-parking-sign.jpg"
      );
      expect(caption).toBeInTheDocument();
    });
  });

  describe("next", () => {
    it("goes to the next image", async () => {
      render(<ImageCarousel images={images} hasAutoScroll={false} />);
      const user = userEvent.setup();

      const nextButton = screen.getByRole("button", {
        name: "Next image",
      });
      await user.click(nextButton);

      const count = screen.getByText("2/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[1].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute("src", "/src/assets/test/lego.jpg");
      expect(caption).toBeInTheDocument();
    });

    it("goes to the first image when on the last image", async () => {
      render(<ImageCarousel images={images} hasAutoScroll={false} />);
      const user = userEvent.setup();
      const thirdImageDot = screen.getByRole("button", {
        name: "Show image 3",
      });
      await user.click(thirdImageDot);

      const nextButton = screen.getByRole("button", {
        name: "Next image",
      });
      await user.click(nextButton);

      const count = screen.getByText("1/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[0].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute(
        "src",
        "/src/assets/test/la-rocher-de-baume.jpg"
      );
      expect(caption).toBeInTheDocument();
    });
  });

  describe("dots", () => {
    it("clicking on dot goes to that nth image", async () => {
      render(<ImageCarousel images={images} hasAutoScroll={false} />);
      const user = userEvent.setup();

      const secondImageDot = screen.getByRole("button", {
        name: "Show image 2",
      });
      await user.click(secondImageDot);

      const count = screen.getByText("2/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[1].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute("src", "/src/assets/test/lego.jpg");
      expect(caption).toBeInTheDocument();
    });

    it("clicking on the first dot goes to the first image", async () => {
      render(<ImageCarousel images={images} hasAutoScroll={false} />);
      const user = userEvent.setup();
      const thirdImageDot = screen.getByRole("button", {
        name: "Show image 3",
      });
      await user.click(thirdImageDot);

      const firstImageDot = screen.getByRole("button", {
        name: "Show image 1",
      });
      await user.click(firstImageDot);

      const count = screen.getByText("1/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[0].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute(
        "src",
        "/src/assets/test/la-rocher-de-baume.jpg"
      );
      expect(caption).toBeInTheDocument();
    });

    it("clicking on the last dot goes to the last image", async () => {
      render(<ImageCarousel images={images} hasAutoScroll={false} />);
      const user = userEvent.setup();

      const thirdImageDot = screen.getByRole("button", {
        name: "Show image 3",
      });
      await user.click(thirdImageDot);

      const count = screen.getByText("3/3");
      const image = screen.getByRole("presentation");
      const caption = screen.getByText(images[2].caption);
      expect(count).toBeInTheDocument();
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute(
        "src",
        "/src/assets/test/self-parking-sign.jpg"
      );
      expect(caption).toBeInTheDocument();
    });
  });
});
