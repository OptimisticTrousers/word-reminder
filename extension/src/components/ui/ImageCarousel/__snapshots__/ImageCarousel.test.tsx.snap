// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageCarousel component > shows no images placeholder when there are no images 1`] = `[Function]`;

exports[`ImageCarousel component > shows the first image by default 1`] = `
<DocumentFragment>
  <div
    class="_carousel_02d13e"
  >
    <div
      class="_carousel__images_02d13e"
    >
      <figure
        class="_carousel__item_02d13e _fade_02d13e"
      >
        <span
          class="_carousel__count_02d13e"
        >
          1/3
        </span>
        <img
          alt=""
          class="_carousel__image_02d13e"
          role="presentation"
          src="/src/assets/test/la-rocher-de-baume.jpg"
        />
        <figcaption
          class="_carousel__caption_02d13e"
        >
          La Rocher De La Baume
        </figcaption>
      </figure>
      <button
        aria-label="Previous image"
        class="_carousel__button_02d13e _carousel__button--prev_02d13e"
      >
        <svg
          class="lucide lucide-chevron-left _carousel__icon_02d13e"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m15 18-6-6 6-6"
          />
        </svg>
      </button>
      <button
        aria-label="Next image"
        class="_carousel__button_02d13e _carousel__button--next_02d13e"
      >
        <svg
          class="lucide lucide-chevron-right _carousel__icon_02d13e"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </button>
    </div>
    <div
      class="_carousel__dots_02d13e"
    >
      <button
        aria-label="Show image 1"
        class="_carousel__button_02d13e _carousel__button--dot_02d13e _carousel__button--active_02d13e"
      />
      <button
        aria-label="Show image 2"
        class="_carousel__button_02d13e _carousel__button--dot_02d13e _false_02d13e"
      />
      <button
        aria-label="Show image 3"
        class="_carousel__button_02d13e _carousel__button--dot_02d13e _false_02d13e"
      />
    </div>
  </div>
</DocumentFragment>
`;
